apps/prairielearn/src/sync/sharing.sql

```diff
@@ -17,3 +18,42 @@ WHERE
         question_id = q.id
     )
   );
+
+-- BLOCK select_sharing_sets
+SELECT
+  ss.name
+FROM
+  sharing_sets AS ss
+WHERE
+  ss.course_id = $course_id;
+
+-- BLOCK select_question_sharing_sets
+SELECT
+  *
+FROM
+  (
+    SELECT
+      q.id,
+      q.qid,
+      q.shared_publicly,
+      (
+        SELECT
+          jsonb_agg(ss.name)
+        FROM
+          sharing_set_questions AS ssq
+          JOIN sharing_sets AS ss on (ss.id = ssq.sharing_set_id)
+        WHERE
+          ssq.question_id = q.id
+      ) AS sharing_sets
+    FROM
+      questions AS q
+    WHERE
+      q.course_id = $course_id
+      AND q.deleted_at IS NULL
+    GROUP BY
+      q.id
+    ORDER BY
+      q.qid
+  ) AS qss
+WHERE
+  sharing_sets IS NOT NULL;
```

Comment:
I haven't looked at this too closely, but is there any way to optimize this? Here's a query plan when run against Mariana's course in production:

```
                                                                            QUERY PLAN                                                                             
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=37269.28..37273.60 rows=1731 width=70) (actual time=62.094..62.099 rows=1 loops=1)
   Sort Key: q.qid
   Sort Method: quicksort  Memory: 25kB
   ->  HashAggregate  (cost=6988.84..37176.17 rows=1731 width=70) (actual time=12.398..62.092 rows=1 loops=1)
         Group Key: q.id
         Filter: ((SubPlan 2) IS NOT NULL)
         Batches: 1  Memory Usage: 321kB
         Rows Removed by Filter: 1657
         ->  Bitmap Heap Scan on questions q  (cost=100.19..6984.49 rows=1740 width=38) (actual time=0.254..2.027 rows=1658 loops=1)
               Recheck Cond: (course_id = 114)
               Filter: (deleted_at IS NULL)
               Rows Removed by Filter: 87
               Heap Blocks: exact=576
               ->  Bitmap Index Scan on questions_course_id_shared_publicly_idx  (cost=0.00..99.76 rows=2045 width=0) (actual time=0.168..0.168 rows=1745 loops=1)
                     Index Cond: (course_id = 114)
         SubPlan 1
           ->  Aggregate  (cost=8.68..8.69 rows=1 width=32) (actual time=0.051..0.054 rows=1 loops=1)
                 ->  Hash Join  (cost=7.33..8.68 rows=1 width=32) (actual time=0.042..0.050 rows=1 loops=1)
                       Hash Cond: (ss.id = ssq.sharing_set_id)
                       ->  Seq Scan on sharing_sets ss  (cost=0.00..1.27 rows=27 width=40) (actual time=0.002..0.005 rows=28 loops=1)
                       ->  Hash  (cost=7.31..7.31 rows=1 width=8) (actual time=0.036..0.037 rows=1 loops=1)
                             Buckets: 1024  Batches: 1  Memory Usage: 9kB
                             ->  Seq Scan on sharing_set_questions ssq  (cost=0.00..7.31 rows=1 width=8) (actual time=0.003..0.034 rows=1 loops=1)
                                   Filter: (question_id = q.id)
                                   Rows Removed by Filter: 443
         SubPlan 2
           ->  Aggregate  (cost=8.68..8.69 rows=1 width=32) (actual time=0.035..0.035 rows=1 loops=1658)
                 ->  Hash Join  (cost=7.33..8.68 rows=1 width=32) (actual time=0.035..0.035 rows=0 loops=1658)
                       Hash Cond: (ss_1.id = ssq_1.sharing_set_id)
                       ->  Seq Scan on sharing_sets ss_1  (cost=0.00..1.27 rows=27 width=40) (actual time=0.004..0.005 rows=14 loops=2)
                       ->  Hash  (cost=7.31..7.31 rows=1 width=8) (actual time=0.034..0.034 rows=0 loops=1658)
                             Buckets: 1024  Batches: 1  Memory Usage: 9kB
                             ->  Seq Scan on sharing_set_questions ssq_1  (cost=0.00..7.31 rows=1 width=8) (actual time=0.034..0.034 rows=0 loops=1658)
                                   Filter: (question_id = q.id)
                                   Rows Removed by Filter: 444
 Planning Time: 0.327 ms
 Execution Time: 62.376 ms
(37 rows)
```

60ms isn't terrible but it does eat into our syncing budget. For comparison, selecting just all the questions takes 10ms.

Reply:
I pulled out the subquery to a CTE (with help from ChatGPT, which is really good at this, by the way). 

```
                                                           QUERY PLAN                                               
            
--------------------------------------------------------------------------------------------------------------------
------------
 Sort  (cost=77.76..77.76 rows=1 width=73)
   Sort Key: q.qid
   ->  Hash Join  (cost=72.72..77.75 rows=1 width=73)
         Hash Cond: (ssq.question_id = q.id)
         ->  HashAggregate  (cost=64.57..67.07 rows=199 width=40)
               Group Key: ssq.question_id
               Filter: (jsonb_agg(ss.name) IS NOT NULL)
               ->  Hash Join  (cost=26.88..56.72 rows=1570 width=40)
                     Hash Cond: (ssq.sharing_set_id = ss.id)
                     ->  Seq Scan on sharing_set_questions ssq  (cost=0.00..25.70 rows=1570 width=16)
                     ->  Hash  (cost=17.50..17.50 rows=750 width=40)
                           ->  Seq Scan on sharing_sets ss  (cost=0.00..17.50 rows=750 width=40)
         ->  Hash  (cost=8.14..8.14 rows=1 width=41)
               ->  Index Scan using questions_course_id_number_nondeleted_key on questions q  (cost=0.12..8.14 rows=
1 width=41)
                     Index Cond: (course_id = 1)
(15 rows)
```

Reply:
Seems much better! Let's continue conversation on https://github.com/PrairieLearn/PrairieLearn/pull/10029#discussion_r1792620053.